akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala

```diff
@@ -885,6 +951,60 @@ abstract class PersistentActorSpec(config: Config) extends PersistenceSpec(confi
 
       expectNoMsg(100.millis)
     }
+    "allow nested persist calls" in {
+      val persistentActor = system.actorOf(Props(classOf[NestedPersists], name, testActor))
+      persistentActor ! "a"
+      persistentActor ! "b"
+
+      expectMsg("a")
+      expectMsg("a-outer")
+      expectMsg("a-inner")
+      // and only then process "b"
+      expectMsg("b")
+      expectMsg("b-outer")
+      expectMsg("b-inner")
+    }
+    "allow multiple persists with nested persist calls" in {
+      val persistentActor = system.actorOf(Props(classOf[MultipleAndNestedPersists], name, testActor))
+      persistentActor ! "a"
+      persistentActor ! "b"
+
+      expectMsg("a")
+      expectMsg("a-outer-1")
+      expectMsg("a-outer-2")
+      expectMsg("a-inner-1")
+      expectMsg("a-inner-2")
+      // and only then process "b"
+      expectMsg("b")
+      expectMsg("b-outer-1")
+      expectMsg("b-outer-2")
+      expectMsg("b-inner-1")
+      expectMsg("b-inner-2")
+    }
+    "allow deeply nested persist calls" in {
+      val nestedPersists = 6
+
+      val persistentActor = system.actorOf(Props(classOf[DeeplyNestedPersists], name, nestedPersists, testActor))
+      persistentActor ! "a"
+      persistentActor ! "b"
+
+      expectMsg("a")
+      receiveN(6) should ===((1 to nestedPersists).map("a-" + _))
+      // and only then process "b"
+      expectMsg("b")
+      receiveN(6) should ===((1 to nestedPersists).map("b-" + _))
+    }
+    "not allow nesting persistAsync calls" in {
+      // currently disalow (loudly) calling persistAsync inside callbacks
+      // TODO rethink it again if it wouldn't be doable... may make sense if persisted data is CRDTs?
```

Comment:
the callbacks didn't get called, did not dig deep enough - will investigate more now